libs/geo/src/main/java/org/opensearch/geometry/utils/WellKnownText.java

```diff
@@ -301,16 +302,95 @@ private Geometry parseGeometry(StreamTokenizer stream) throws IOException, Parse
         throw new IllegalArgumentException("Unknown geometry type: " + type);
     }
 
+    /**
+     * Iterative version of
+     * <!--
+     * ```java
+     *     private GeometryCollection<Geometry> parseGeometryCollectionA(StreamTokenizer stream) throws IOException, ParseException {
+     *         if (nextEmptyOrOpen(stream).equals(EMPTY)) {
+     *             return GeometryCollection.EMPTY;
+     *         }
+     *         List<Geometry> shapes = new ArrayList<>();
+     *         shapes.add(parseGeometry(stream));
+     *         while (nextCloserOrComma(stream).equals(COMMA)) {
+     *             shapes.add(parseGeometry(stream));
+     *         }
+     *         return new GeometryCollection<>(shapes);
+     *     }
+     * -->
+     * to avoid StackOverflowError when there is a deeply nested structure of GeometryCollection.
+     */
     private GeometryCollection<Geometry> parseGeometryCollection(StreamTokenizer stream) throws IOException, ParseException {
         if (nextEmptyOrOpen(stream).equals(EMPTY)) {
             return GeometryCollection.EMPTY;
         }
-        List<Geometry> shapes = new ArrayList<>();
-        shapes.add(parseGeometry(stream));
-        while (nextCloserOrComma(stream).equals(COMMA)) {
-            shapes.add(parseGeometry(stream));
+
+        List<Geometry> topLevelShapes = new ArrayList<>();
+        Stack<List<Geometry>> stack = new Stack<>();
+        stack.push(topLevelShapes);
+        boolean isFirstIteration = true;
+        List<Geometry> currentLevelShapes = null;
+        while (!stack.isEmpty()) {
+            List<Geometry> previousShapes = stack.pop();
+            if (currentLevelShapes != null) {
+                previousShapes.add(new GeometryCollection<>(currentLevelShapes));
+            }
+            currentLevelShapes = previousShapes;
+
+            if (isFirstIteration == true) {
+                isFirstIteration = false;
+            } else {
+                if (!nextCloserOrComma(stream).equals(COMMA)) {
+                    continue;
```

Comment:
This `continue` relies on that fact that you did `pop` earlier. 

Maybe add a comment like "Done with current level, continue with parent level".